src/librustc/traits/project.rs

```diff
@@ -411,8 +411,9 @@ impl<'a, 'b, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for AssociatedTypeNormalizer<'a,
                         };
                         if let Ok(evaluated) = tcx.const_eval(param_env.and(cid)) {
                             let substs = tcx.lift_to_global(&substs).unwrap();
```

Comment:
Do `substs` need to be lifted here?  If `substs` needed inference but the const still evaluated successfully, doesn't that mean we can just use `InternalSubsts::empty()` here?

Reply:
Or am I misunderstanding what's happening here?

Reply:
This is confusing - `InternalSubsts::empty()` would be wrong, but I'm wondering why there is any substitution at all, maybe this is from the time when this could return e.g. function pointers referring to generic parameters?

IMO `const_eval` should either return an (interned) `&'tcx ty::Const`, or the `Scalar | Slice | ByRef` enum by value, not what it does now.

Reply:
Oh, nevermind, `const_eval` returns the type as well, which *could* depend on parameters and whatnot (especially lifetime ones!).

Reply:
Ah, cool, makes sense. The reason I ask is that this blows up on array lengths that use generic parameters when they have a real param_env.